src/rcheevos/memref.c

```diff
@@ -0,0 +1,204 @@
+
+  return memref_value;
+}
+
+static unsigned rc_memref_get_value(rc_memref_t* self, rc_peek_t peek, void* ud) {
+  unsigned value;
+
+  switch (self->size)
+  {
+    case RC_MEMSIZE_BIT_0:
+      value = (peek(self->address, 1, ud) >> 0) & 1;
+      break;
+
+    case RC_MEMSIZE_BIT_1:
+      value = (peek(self->address, 1, ud) >> 1) & 1;
+      break;
+
+    case RC_MEMSIZE_BIT_2:
+      value = (peek(self->address, 1, ud) >> 2) & 1;
+      break;
+
+    case RC_MEMSIZE_BIT_3:
+      value = (peek(self->address, 1, ud) >> 3) & 1;
+      break;
+
+    case RC_MEMSIZE_BIT_4:
+      value = (peek(self->address, 1, ud) >> 4) & 1;
+      break;
+
+    case RC_MEMSIZE_BIT_5:
+      value = (peek(self->address, 1, ud) >> 5) & 1;
+      break;
+
+    case RC_MEMSIZE_BIT_6:
+      value = (peek(self->address, 1, ud) >> 6) & 1;
+      break;
+
+    case RC_MEMSIZE_BIT_7:
+      value = (peek(self->address, 1, ud) >> 7) & 1;
+      break;
+
+    case RC_MEMSIZE_LOW:
+      value = peek(self->address, 1, ud) & 0x0f;
+      break;
+
+    case RC_MEMSIZE_HIGH:
+      value = (peek(self->address, 1, ud) >> 4) & 0x0f;
+      break;
+
+    case RC_MEMSIZE_8_BITS:
+      value = peek(self->address, 1, ud);
+
+      if (self->is_bcd) {
+         value = ((value >> 4) & 0x0f) * 10 + (value & 0x0f);
+      }
+
+      break;
+
+    case RC_MEMSIZE_16_BITS:
+      value = peek(self->address, 2, ud);
+
+      if (self->is_bcd) {
+         value = ((value >> 12) & 0x0f) * 1000
+               + ((value >> 8) & 0x0f) * 100
+               + ((value >> 4) & 0x0f) * 10
+               + ((value >> 0) & 0x0f) * 1;
+      }
+
+      break;
+
+    case RC_MEMSIZE_24_BITS:
+      value = peek(self->address, 4, ud);
+
+      if (self->is_bcd) {
+        value = ((value >> 20) & 0x0f) * 100000
+              + ((value >> 16) & 0x0f) * 10000
+              + ((value >> 12) & 0x0f) * 1000
+              + ((value >> 8) & 0x0f) * 100
+              + ((value >> 4) & 0x0f) * 10
+              + ((value >> 0) & 0x0f) * 1;
+      }
+
+      break;
+
+    case RC_MEMSIZE_32_BITS:
+      value = peek(self->address, 4, ud);
+
+      if (self->is_bcd) {
+        value = ((value >> 28) & 0x0f) * 10000000
+              + ((value >> 24) & 0x0f) * 1000000
+              + ((value >> 20) & 0x0f) * 100000
+              + ((value >> 16) & 0x0f) * 10000
+              + ((value >> 12) & 0x0f) * 1000
+              + ((value >> 8) & 0x0f) * 100
+              + ((value >> 4) & 0x0f) * 10
+              + ((value >> 0) & 0x0f) * 1;
+      }
+
+      break;
+
+    default:
+      value = 0;
+      break;
+  }
+
+  return value;
+}
+
+void rc_update_memref_values(rc_memref_value_t* memref, rc_peek_t peek, void* ud) {
+  while (memref) {
+    memref->previous = memref->value;
+    memref->value = rc_memref_get_value(&memref->memref, peek, ud);
+    if (memref->value != memref->previous)
+      memref->prior = memref->previous;
+
+    memref = memref->next;
+  }
```

Comment:
memref processing loop traverses chain and updates the three data values appropriately